server/datastore/mysql/hosts.go

```diff
@@ -907,9 +946,45 @@ func (d *Datastore) CleanupExpiredHosts(ctx context.Context) error {
 	if !ac.HostExpirySettings.HostExpiryEnabled {
 		return nil
 	}
-	_, err = d.writer.ExecContext(ctx, `DELETE FROM hosts WHERE seen_time < DATE_SUB(NOW(), INTERVAL ? DAY)`, ac.HostExpirySettings.HostExpiryWindow)
+
+	// Usual clean up queries used to be like this:
+	// DELETE FROM hosts WHERE id in (SELECT host_id FROM host_seen_times WHERE seen_time < DATE_SUB(NOW(), INTERVAL ? DAY))
+	// This means a full table scan for hosts, and for big deployments, that's not ideal
+	// so instead, we get the ids one by one and delete things one by one
+	// it might take longer, but it should lock only the row we need
+
+	rows, err := d.writer.QueryContext(
+		ctx,
+		`SELECT host_id FROM host_seen_times WHERE seen_time < DATE_SUB(NOW(), INTERVAL ? DAY)`,
+		ac.HostExpirySettings.HostExpiryWindow,
+	)
+	if err != nil {
+		return errors.Wrap(err, "getting expired host ids")
+	}
+	defer rows.Close()
+
+	for rows.Next() {
```

Comment:
<Just a nit, as this PR has been up for too long now :)>

Small improvement to not mix row scanning + deletion in the same for loop: 

1. Make a helper function that calls the `SELECT` and returns a slice of ids (the `rows.Err() + rows.Close()` would get call when the function returns).
2. Then on the for loop here you iterate such returned slice of `id`s and call `DELETE` on them.

Reply:
The reasoning behind the scan+delete in a loop is to not store potentially a silly amount of ids in memory, though.